tests/unit/test_charm.py

```diff
@@ -85,59 +97,124 @@ def test_update_container_correct_pebble_layer(
+def test_on_database_created_cannot_connect_container(
+    harness, mocked_kubernetes_service_patcher
+) -> None:
     harness.begin()
-    harness.set_can_connect(CONTAINER_NAME, True)
+    harness.set_can_connect(CONTAINER_NAME, False)
 
-    container = harness.model.unit.get_container(CONTAINER_NAME)
-    harness.charm.on.kratos_pebble_ready.emit(container)
+    setup_postgres_relation(harness)
 
-    assert isinstance(harness.charm.unit.status, BlockedStatus)
+    assert isinstance(harness.charm.unit.status, WaitingStatus)
 
 
-def test_database_not_created(harness, mocked_kubernetes_service_patcher) -> None:
+def test_on_database_created_before_pebble_ready(
+    harness, mocked_kubernetes_service_patcher, mocked_sql_migration
+) -> None:
     harness.begin()
     harness.set_can_connect(CONTAINER_NAME, True)
+    setup_postgres_relation(harness)
+
+    assert isinstance(harness.charm.unit.status, WaitingStatus)
+    mocked_sql_migration.assert_not_called()
 
-    db_relation_id = harness.add_relation("pg-database", "postgresql-k8s")
-    harness.add_relation_unit(db_relation_id, "postgresql-k8s/0")
+
+def test_on_database_created_after_pebble_ready(
+    harness, mocked_kubernetes_service_patcher, mocked_sql_migration
+) -> None:
+    harness.begin()
+    harness.set_can_connect(CONTAINER_NAME, True)
 
     container = harness.model.unit.get_container(CONTAINER_NAME)
     harness.charm.on.kratos_pebble_ready.emit(container)
 
-    assert isinstance(harness.charm.unit.status, WaitingStatus)
+    setup_postgres_relation(harness)
+
+    service = harness.model.unit.get_container("kratos").get_service("kratos")
+    assert service.is_running()
+    assert isinstance(harness.charm.unit.status, ActiveStatus)
+    mocked_sql_migration.assert_called_once()
+
+    updated_config = yaml.safe_load(harness.charm._config)
+    assert DB_ENDPOINTS in updated_config["dsn"]
+    assert DB_PASSWORD in updated_config["dsn"]
+    assert DB_USERNAME in updated_config["dsn"]
 
 
-def test_on_pebble_ready(
-    harness, mocked_kubernetes_service_patcher, mocked_update_container
+def test_on_database_changed_cannot_connect_container(
+    harness, mocked_kubernetes_service_patcher
 ) -> None:
+    harness.begin()
+    harness.set_can_connect(CONTAINER_NAME, False)
+
+    trigger_database_changed(harness)
+
+    assert isinstance(harness.charm.unit.status, WaitingStatus)
 
+
+def test_on_database_changed_before_pebble_ready(
+    harness, mocked_kubernetes_service_patcher
+) -> None:
     harness.begin()
     harness.set_can_connect(CONTAINER_NAME, True)
-    setup_postgres_relation(harness)
 
-    mocked_update_container.assert_called_once()
+    trigger_database_changed(harness)
+
+    assert isinstance(harness.charm.unit.status, WaitingStatus)
 
 
-def test_on_database_created(
-    harness, mocked_kubernetes_service_patcher, mocked_update_container
+def test_on_database_changed_after_pebble_ready(
```

Comment:
I would recommend using `given_x_when_y_then_z` as a test naming convention. This avoids using words like `with` and `after` and makes it obvious for readers of the unit tests what the expected behaviour is